### PR TITLE
[Feature] support new sampler backend with triton

### PR DESCRIPTION
**Pull request by zeroRains**
## Motivation

为 FastDeploy 添加新的基于 Triton 的采样后端,提升 Top-K/Top-P 采样的性能。现有的采样实现存在以下问题：
1. 采样操作涉及多次 CPU-GPU 同步，带来显著的延迟开销
2. Top-K/Top-P 采样算法效率不够高，尤其在大词表场景下
3. 缺少高效的 GPU 端随机数生成机制

本 PR 引入 vLLM 的 Triton 采样后端（基于论文 "Qrita: High-performance Top-k and Top-p Algorithm for GPUs using Pivot-based Truncation and Selection"），通过以下优化提升采样性能：
- 使用 Triton 内核实现高效的 Top-K/Top-P 联合采样
- 在 Softmax 之前直接对 logits 进行 mask，避免概率空间的转换开销
- 使用 Philox PRNG 在 GPU 端生成确定性的 Gumbel 噪声，消除 CPU-GPU 同步
- 基于枢轴的截断和选择算法，显著降低大词表场景的计算复杂度

## Modifications

本 PR 共修改 **4 个文件**，新增 1127 行，删除 30 行：

### 1. **新增文件：`fastdeploy/model_executor/layers/sample/ops/top_k_top_p_triton.py`** (+987 行)

**核心功能**：实现基于 Triton 的 Top-K/Top-P 联合采样内核

**主要组件**：

- **`_topk_topp_kernel`**: Triton JIT 内核，实现联合 Top-K/Top-P 采样
  - 支持三种模式：Top-K only、Top-P only、Top-K + Top-P
  - 使用三元搜索（Ternary Search）快速定位 pivot
  - 基于 Gaussian sigma-truncation 进行离群值收集，减少搜索空间
  - 支持返回采样 mask（零额外带宽成本）

- **`apply_top_k_top_p_triton`**: Python 接口函数
  - 输入：`[batch_size, vocab_size]` 的 logits 张量
  - 参数：`k` (Top-K 值)、`p` (Top-P 值)、`mask_value` (mask 填充值)
  - 输出：in-place 修改后的 logits，可选返回 bool mask

- **`seeded_gumbel_noise`**: GPU 端 Gumbel 噪声生成
  - 使用 Philox PRNG 生成确定性随机数
  - 支持 per-request seed，确保可复现性
  - 完全在 GPU 端执行，无 CPU-GPU 同步开销

- **查找表**：
  - `_NORMAL_CDF_TO_SIGMA_TABLE`: 正态分布 CDF 到 sigma 的映射
  - `_PERCENTILE_TO_STD_TABLE`: 百分位到标准差的映射
  - 用于快速计算采样所需的统计量

- **缓存机制**：
  - `_TRITON_BUFFER_CACHE`: 缓存 per-program 工作缓冲区
  - `_TRITON_TABLE_CACHE`: 缓存查找表张量
  - 避免重复分配内存，降低开销

### 2. **修改文件：`fastdeploy/model_executor/layers/sample/sampler.py`** (+138 行, -29 行)

**主要改动**：

- **新增函数 `_apply_triton_top_k_top_p`**：
  - 封装 Triton Top-K/Top-P mask 操作
  - 在 Softmax **之前**对 logits 进行 mask（关键优化点）
  - 支持返回 mask 张量（用于 Keep Sampling Mask 功能）

- **新增函数 `_random_sample`**：
  - 使用 Gumbel-max trick 实现 multinomial 采样
  - 当提供 `topp_seed` 时，调用 Triton 内核生成确定性噪声
  - 避免原有的 Python for-loop 和 CPU-GPU 同步

- **修改采样流程**（3 处调用点）：
  - **`forward_cuda` (正常采样路径)**：
    ```python
    if FD_SAMPLING_CLASS.lower() == "triton":
        logits = _apply_triton_top_k_top_p(logits, top_p, top_k, ...)
        probs = F.softmax(logits)
        next_tokens = _random_sample(probs, topp_seed=seed)
    else:
        probs = F.softmax(logits)
        _, next_tokens = top_k_top_p_sampling(probs, ...)
    ```

  - **`_verify_and_sample` (投机解码路径)**：
    - 在 `TARGET_MATCH` 策略下使用 Triton 采样
    - 保持与原有采样逻辑的兼容性

  - **`_normal_sample` (MTP 普通采样路径)**：
    - 支持 Triton 采样后端
    - 与多令牌预测（MTP）功能兼容

- **Logprobs 计算兼容性**：
  - 当使用 Triton 采样时，保存原始 logits (`logits_ori`)
  - 确保 logprobs 基于未 mask 的原始 logits 计算
  - 适配 `forward_cuda` 中的 `build_output_logprobs` 调用

### 3. **修改文件：`fastdeploy/envs.py`** (+1 行, -1 行)

**改动说明**：
- 更新环境变量 `FD_SAMPLING_CLASS` 的注释文档
- 原文档：`"base", "base_non_truncated", "air" and "rejection" can be set currently`
- 新文档：`"base", "base_non_truncated", "air", "rejection" and "triton" can be set currently`
- 添加了 `"triton"` 选项，用于启用 Triton 采样后端

### 4. **修改文件：`.flake8`** (+1 行)

**改动说明**：
- 新增对 `top_k_top_p_triton.py` 文件的 flake8 规则豁免
- 豁免规则：`E241` (多余空格), `E121` (缩进), `E131` (缩进), `E266` (注释格式)
- 原因：Triton 内核代码格式与 PEP8 存在差异，需要保持与 vLLM 上游代码的一致性

### 启用 Triton 采样后端

通过环境变量 `FD_SAMPLING_CLASS` 启用：

```bash
MODEL_PATH="/root/paddlejob/tmpspace/GLM-4.5-Air-New/"
export FD_SAMPLING_CLASS=triton
python -m fastdeploy.entrypoints.openai.api_server \
    --port 9513 \
    --host $(hostname -i) \
    --model "$MODEL_PATH" \
    --disable-custom-all-reduce \
    --tensor-parallel-size 4 \
    --max-model-len 131072 \
    --max-num-seqs 32 \
    --gpu-memory-utilization 0.9 \
    --graph-optimization-config '{"use_cudagraph":true}' \
    --enable-logprob \
    --speculative-config '{"method":"mtp","num_speculative_tokens":1,"num_model_steps":1,"model":"'$MODEL_PATH'"}' \
```
### 请求
```bash
curl -X POST "http://localhost:9513/v1/chat/completions" -H "Content-Type: application/json" -d '{
    "messages": [
        {"role": "system", "content": "你是谁"}
    ] ,
    "temperature":1.0,
    "top_p": 0,
    "seed": 2333,
    "top_k": 1,
    "logprobs": true,
    "top_logprobs": 0
  }'
```

## Accuracy Tests

输出结果无乱码，logrpobs计算正常。

前半段计算一致，后半段计算不同，随机性不一致。

image

Coverage说明：新增单测可以测到修改内容，但triton算子执行时变成ptx了，coverage监测不到

## Checklist

- [x] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [x] Format your code, run `pre-commit` before commit.
- [x] Add unit tests. Please write the reason in this PR if no unit tests.
- [x] Provide accuracy results.
- [x] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

**Review by yuanlehome — APPROVED** [review]
LGTM
